crates/core/src/repofile/snapshotfile.rs

```diff
@@ -1441,4 +1509,166 @@ mod tests {
+            vec![
+                (
+                    id1,
+                    DateTime::from_timestamp(1_752_483_600, 0).unwrap().into(),
+                ),
+                (
+                    id2,
+                    DateTime::from_timestamp(1_752_483_700, 0).unwrap().into(),
+                ),
+                (
+                    // this is the latest
+                    id3,
+                    DateTime::from_timestamp(1_752_483_800, 0).unwrap().into(),
+                ),
+            ],
+            &key,
+        );
+
+        let mut back = MockBackend::new();
+        let _ = back.expect_list_with_size().returning(move |_| {
+            // unordered ids
+            Ok(vec![(id2, 0), (id3, 0), (id1, 0)])
+        });
+        let _ = back
+            .expect_read_full()
+            .returning(move |_tpe, id| Ok(snapshot_files.get(id).unwrap().clone()));
+
+        let mut be = DecryptBackend::new(Arc::new(back), key);
+        be.set_zstd(None);
+
+        let latest = SnapshotFile::latest(&be, |_sn| true, &NoProgress).unwrap();
+        assert_eq!(latest.id, SnapshotId(id3));
+
+        let latest_n0 = SnapshotFile::latest_n(&be, |_sn| true, &NoProgress, 0).unwrap();
+        assert_eq!(latest_n0, latest);
+
+        let latest_n1 = SnapshotFile::latest_n(&be, |_sn| true, &NoProgress, 1).unwrap();
+        assert_eq!(latest_n1.id, SnapshotId(id2));
+
+        let latest_n2 = SnapshotFile::latest_n(&be, |_sn| true, &NoProgress, 2).unwrap();
+        assert_eq!(latest_n2.id, SnapshotId(id1));
+
+        let latest_n3 = SnapshotFile::latest_n(&be, |_sn| true, &NoProgress, 3);
+        let latest_n3_err = latest_n3.unwrap_err().to_string();
+        let expected = "No snapshots found for latest~3.";
+        assert!(
+            latest_n3_err.contains(expected),
+            "Err is: {latest_n3_err}\n\nShould contain: {expected}",
+        );
+    }
+
+    #[rstest]
+    fn test_snapshot_file_from_str() {
+        let key = Key::new();
+        let id1 = Id::from_str("0011223344556677001122334455667700112233445566770000000000000001")
+            .unwrap();
+        let id2 = Id::from_str("0011223344556677001122334455667700112233445566770000000000000002")
+            .unwrap();
+        let id3 = Id::from_str("0011223344556677001122334455667700112233445566770000000000000003")
+            .unwrap();
+
+        let snapshot_files = fake_snapshot_file_with_id_time(
+            vec![
+                (
+                    id1,
+                    DateTime::from_timestamp(1_752_483_600, 0).unwrap().into(),
+                ),
+                (
+                    id2,
+                    DateTime::from_timestamp(1_752_483_700, 0).unwrap().into(),
+                ),
+                (
+                    // this is the latest
+                    id3,
+                    DateTime::from_timestamp(1_752_483_800, 0).unwrap().into(),
+                ),
+            ],
+            &key,
+        );
+
```

Comment:
this is duplicate to the above code. Can you make a single set-up func for this?